### PR TITLE
Feature/uploader

### PR DESCRIPTION
**Pull request by bwalsh**
This PR:
* adds `projects ping`, `projects ls`, `meta import dir` and `files cp file bucket`
* Adds multi-processing (but single-part) file uploader.

```
$ gen3_util projects ping
msg: OK connected to endpoint https://aced-training.compbio.ohsu.edu

$ gen3_util projects ls
endpoint: https://aced-training.compbio.ohsu.edu
projects:
- /programs/aced/projects/Alcoholism
- /programs/aced/projects/Alzheimers
- /programs/aced/projects/Breast_Cancer
- /programs/aced/projects/Colon_Cancer
- /programs/aced/projects/Diabetes
- /programs/aced/projects/Lung_Cancer
- /programs/aced/projects/Prostate_Cancer
buckets:
  GS_BUCKETS: {}
  S3_BUCKETS:
    aced-default:
      endpoint_url: https://minio-default.compbio.ohsu.edu
      region: us-east-1
    aced-manchester:
      endpoint_url: https://minio-manchester.compbio.ohsu.edu
      region: us-east-1
    aced-ohsu:
      endpoint_url: https://minio-ohsu.compbio.ohsu.edu
      region: us-east-1
    aced-stanford:
      endpoint_url: https://minio-stanford.compbio.ohsu.edu
      region: us-east-1
    aced-ucl:
      endpoint_url: https://minio-ucl.compbio.ohsu.edu
      region: us-east-1


$ gen3_util meta  import dir tests/fixtures/dir_to_study/ tmp/foo --project_id aced-Alcoholism
msg: OK
summary:
  DocumentReference:
    count: 5
    size: 6013814
  ResearchStudy:
    count: 1

$ gen3_util files cp  --project_id aced-Alcoholism tmp/foo/DocumentReference.ndjson  bucket://aced-ohsu
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████| 5.74M/5.74M [00:03<00:00, 1.71MB/s, elapsed=0:00:02.056022, file=6f8101]
errors: []
incomplete: []
info:
- Wrote state to ~/.gen3/gen3-util-state/state.ndjson
msg: OK


```

### CODE REVIEW

File: gen3_util/cli/__init__.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+import click
+
+ENV_VARIABLE_PREFIX = 'GEN3_UTIL'
+
+
+def _common_options(self):
+    """Insert common commands into Group or Command."""
+    assert len(self.params) == 0
+    self.params.insert(0,
```
Comment (lbeckman314) [inline gen3_util/cli/__init__.py:15]: I haven't encountered this pattern before --
- Is `self.params` a List?
- Can any additional click Options be added by following the same `insert(n+1...` pattern?

File: gen3_util/cli/__init__.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        super().__init__(*args, **kwargs)
+        _common_options(self)
+
+    def list_commands(self, ctx):
+        """Commands in order of appearance"""
+        return self.commands.keys()
+
+
+class CLIOutput:
+    """Ensure output, exceptions and exit code are returned to user consistently."""
```
Comment (lbeckman314) [inline gen3_util/cli/__init__.py:80]: Fully support consistent user output behavior! Is this a Click-specific change (something to add to future Click projects)?

File: gen3_util/common/__init__.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            ).decode()
+        )
+    else:
+        print(output)
+
+
+def read_ndjson_file(path: str) -> Iterator[dict]:
+    """Read ndjson file, load json line by line."""
+    with _file_opener(path) as jsonfile:
+        for l_ in jsonfile.readlines():
```
Comment (lbeckman314) [inline gen3_util/common/__init__.py:33]: Small syntax question -- what cases would the underscore following a variable name (e.g. `l_` as above) be used?

File: gen3_util/common/__init__.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        """destination directory"""
+        self.emitters = {}
+        """open file pointers"""
+        self.verbose = verbose
+        """log activity"""
+        self.file_mode = file_mode
+        """mode for file opens"""
+        self.logger = logger
+
+    def __enter__(self):
```
Comment (lbeckman314) [inline gen3_util/common/__init__.py:124]: This whole class is great! Will `__enter__` context manager have any additional functionality included with it? [Reading up on it](https://stackoverflow.com/questions/1984325/explaining-pythons-enter-and-exit) to make sure I understand it.

File: gen3_util/files/cli.py [hunk 5]
```diff
@@ -24,20 +25,38 @@ def files_ls(config: Config):
...
+@click.option('--worker_count', default=10, show_default=True,
+              help="Number of worker processes")
+@click.option('--ignore_state', default=False, is_flag=True, show_default=True,
+              help="Upload file, even if already uploaded")
+@click.option('--project_id', default=None, required=True, show_default=True,
+              help="Gen3 program-project")
+@click.option('--source_path', required=False, default=None, show_default=True,
+              help='Path on local file system')
+@click.option('--disable_progress_bar', default=False, is_flag=True, show_default=True,
+              help="Show progress bar")
```
Comment (lbeckman314) [inline gen3_util/files/cli.py:39]: Might "Show progress bar" be "Hide progress bar" if it's describing the `--disable_progress_bar` flag?

File: gen3_util/files/cli.py [hunk 6]
```diff
@@ -24,20 +25,38 @@ def files_ls(config: Config):
...
-def files_cp(config: Config, from_: str, to_: str):
+def files_cp(config: Config, from_: str, to_: str, worker_count: int, ignore_state: bool, project_id: str,
+             source_path: str, disable_progress_bar: bool):
     """Copy files to/from the project bucket.
 
     Args:
-        FROM: Source url or path
-        TO: Destination url or path
+        from_: Source url or path
+        to_: Destination url or path
```
Comment (lbeckman314) [inline gen3_util/files/cli.py:47]: Same syntax question as `l_` above.

File: gen3_util/files/uploader.py [hunk 7]
```diff
@@ -1,18 +1,348 @@
...
+            while True:
+                results_to_remove = []
+                for record in results:
+                    if record.ready() and record.successful():
+                        r = record.get()
+                        if r.exception:
+                            exceptions[r.document_reference['id']] = {
+                                'exception': str(r.exception),
+                                'document_reference': {
+                                    'id': r.document_reference
```
Comment (lbeckman314) [inline gen3_util/files/uploader.py:282]: Adding the document reference to the exception is a great idea.

File: gen3_util/meta/__init__.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import importlib
+import logging
+import pathlib
+import uuid
+from typing import Iterator, Any
+
+# TODO fix me, make configurable
```
Comment (lbeckman314) [inline gen3_util/meta/__init__.py:7]: What component here would be made configurable?

File: gen3_util/projects/__init__.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        auth = ensure_auth(config.gen3.refresh_file)
+
+    response = auth.curl('/user/data/buckets')
+
+    if response.status_code == 405:
+        logging.getLogger(__name__).warning(
+            "TODO /data/buckets response returned 405, "
+            "see https://cdis.slack.com/archives/CDDPLU1NU/p1683566639636949 "
+            "see quay.io/cdis/fence:feature_bucket_info_endpoint "
+        )
```
Comment (lbeckman314) [inline gen3_util/projects/__init__.py:29]: No longer necessary? (as of only a couple hours ago haha -- endpoint added to latest master branch of fence)

File: gen3_util/projects/cli.py [hunk 10]
```diff
@@ -14,11 +15,21 @@ def project_group(config: Config):
     pass
 
 
+@project_group.command(name="ping")
+@click.pass_obj
+def ping(config: Config):
```
Comment (lbeckman314) [inline gen3_util/projects/cli.py:20]: Great idea, very functional and nice to have.

**Review by lbeckman314 — APPROVED** [review]
Really great additions, can see this being used by a number of Gen3 commons. Cloning and running tests now but nothing blocking PR. Excited to have this utility moving forward!

**Comment by matthewpeterkort** [comment]
couldn't get the ping command to work despite having a fresh creds file from my localhost https://aced-training.compbio.
ohsu.edu/identity. populated in ~/.gen3/fence/ and ~/.gen3/

Screen Shot 2023-05-08 at 4 00 48 PM

File: gen3_util/projects/__init__.py (hunk 9 above)
Comment (matthewpeterkort) [reply gen3_util/projects/__init__.py:29]: yes I get a 405 too at: https://aced-training.compbio.ohsu.edu/user/data/buckets with default fence image.

**Pull request merged**